src/cargo/util/interning.rs

```diff
@@ -46,7 +46,14 @@ impl<'a> From<&'a String> for InternedString {
 
 impl From<String> for InternedString {
     fn from(item: String) -> Self {
-        InternedString::new(&item)
+        let mut cache = interned_storage();
+        let s = cache.get(item.as_str()).copied().unwrap_or_else(|| {
+            let s = item.leak();
+            cache.insert(s);
+            s
+        });
+
+        InternedString { inner: s }
```

Comment:
Without performance data, this falls to being a refactor.  My main concern is that we're duplicating the logic for creating an InternedString.

Unsure whether thats sufficient to block this or not.  Will think more on this

Reply:
Another way to handle this is having a `Cow` variant of `new()`, and then both `new()` and `from()` call it.

Reply:
Uncertainty about the additional complexity is why this was not done originally.

If we want to make it part of the type system without duplicating the core logic we could do something like:

```rust
fn new_inner<S: AsRef<str>>(str: S, make_stat: impl Fn(S) -> &'static str) -> InternedString {
    let mut cache = interned_storage();
    let s = cache.get(str.as_ref()).copied().unwrap_or_else(|| {
        let s = make_stat(str);
        cache.insert(s);
        s
    });
    InternedString { inner: s }
}
```

I don't know if it's worth the additional complexity.

Reply:
imo cow offers us a fairly clean design and I don't expect the check will impact performance all that much